index.html

```diff
@@ -956,11 +956,13 @@ <h2>
           follows:
         </p>
         <ol class="algorithm">
+          <li>Let |window:Window| be the [=relevant global object=] of the
+          [=environment settings object/responsible document=].
```

Comment:
Which environment settings object's responsible document is this?

You should use the relevant global object of [this](https://heycam.github.io/webidl/#this) instead. See the general advice in https://html.spec.whatwg.org/multipage/webappapis.html#realms-settings-objects-global-objects

Reply:
To be clear, the choice of global is not a specific problem for user activation, so any guidance does not belong there. It's the same as any other time you choose a global, which is why I directed you to the section that gives guidance on choosing a global.

Was the paragraph saying "you almost always want the relevant global" not clear enough? What could we do to make it clearer?

Reply:
> Was the paragraph saying "you almost always want the relevant global" not clear enough? 

See below about relocating that text, as it's way too easy to miss!  And way too far down from where a reader needs it.

> What could we do to make it clearer?

The warning about not using "Incumbent" and "entry" needs to be moved to where "Incumbent" and "entry" is defined... it's currently sandwiched between two examples, and looks like it was a spec issue and not **really important guidance**. I totally just skipped over that thinking it was non-normative stuff and some random spec issue/discussion. 

Also, imagine, from my perspective, I just want to get in and out of the HTML spec as quickly as possible. I don't want to dive into the realms etc. unless I absolutely need to. I already understood how "triggered by user activation works", and all I care about is "will it consume it?"... rest is interesting but 🤷‍♂️. 

The majority of cases for checking the activation are to prevent usage without activation (and throw and error or whatever) - so clear(er) setup steps are still needed (get the global, do the check). Get in, get out, happy spec author (why I also suggested just adding an IDL extended attribute). 

Ok, so even after this discussion, and rereading stuff, I'm still not sure what I'm supposed to change this too? I'm really interested to see how other spec authors going to phrase this (I'm hopefully not the only thick/confused spec Editor here:)).

Reply:
I don't think there are really any shortcuts here. The "get in and out as quickly as possible" attitude is not tenable, and I'm not OK with making edits to HTML to encourage it. As a spec editor, dealing with global objects is one of the most fundamental things one needs to understand. One needs to read and grok that section fully.

In particular, the latest patch is missing a statement of which object's relevant global you are looking at. This is emphasized three times in the section I pointed you at:

> When writing algorithms, **the most prominent platform object whose relevant Realm might be important is the this value of the currently-running function object.** In some cases, there can be other important relevant Realms, such as those of any arguments.

and

> Note how the entry, incumbent, and current concepts are usable without qualification, whereas the relevant concept must be applied to a particular platform object.

and 

> In general, web platform specifications should use the relevant concept, applied to the object being operated on (**usually the this value of the current method**).

I'm not really sure how to deal with situations where we have three separate sentences attempting to make something clear, but spec editors are attempting to skim and thus missing them... Suggestions appreciated.

I think in the end it may be offputting that you're coming to this understanding via something that used to be simple-but-wrong (user activation), and being forced to grasp the concept of different global objects in order to fix user activation. But fundamentally different global object usage pervades all specs, including this one; we've spent several issue threads previously working on this (e.g. https://github.com/w3c/payment-request/pull/811).

Reply:
>  Suggestions appreciated

I think we should add a diagram showing how all the things relate - there is just a lot of concepts there, which go down deep rabbit holes... It be great if someone could do a conference talk (or similar) explaining all the relationships in a more human way, tbh.

Reply:
> You should use the relevant global object of **this** instead.

Just want to apologize - I thought that by linking to "this" initially you were pointing me to "this section of WebIDL" - not to a `this` value. I'd forgotten that there was a **this** value available via WebIDL 😅. 

It's the old, "I'd make a JS joke about what `this` is, but I just don't know what `this` is." 😬... 🥺. 

Anyway, to finish this off, should I pull the relevant global from `this` or from the "payment-relevant browsing context"?

Reply:
> Anyway, to finish this off, should I pull the relevant global from this or from the "payment-relevant browsing context"?

That's a great question, and really gets at some of the subleties here.

Per https://html.spec.whatwg.org/#activation-notification:

- Whenever an iframed Window is activated, its ancestor (payment-relevant browsing context's Window) will be activated too
- Whenever a payment-relevant browsing context is activated, its descendant will get activation, if its descendant is same-origin.

So the question for the spec editors is, do you want to allow clicking on the top-level Window to activate a PaymentRequest object that comes from a same-origin iframe? (You will never be able to allow clicking on the top-level Window to activate a PaymentRequest object that comes from a different-origin iframe, within the current framework.)

If you do, then checking the payment-relevant browsing context's WindowProxy's [[Window]] makes sense. If you don't, then checking the relevant global of this makes sense.

Personally I'd lean toward the more conservative option, but I'm unsure what folks have implemented, or what kind of workflows the PaymentRequest API hopes to support in this fashion.

Reply:
> You will never be able to allow clicking on the top-level Window to activate a PaymentRequest object that comes from a different-origin iframe, within the current framework.

Oh, um, ok... we might need to check how Stripe and Shopify work (why we have the “allowpaymentrequest” on iframes). That might be a showstopper if they currently depend on this behavior. 

Did the above used to work for other things in old v1 activation model? I seem to recall “triggered by user activation” needing to survive a postMessage() to work with some payment services.

Reply:
I suspect they work by clicking within the iframe (but I don't know).

Reply:
Ok, Stripe depends on clicking within the iframe - so we are good there. 

![Screenshot 2019-12-11 12 48 17](https://user-images.githubusercontent.com/870154/70584222-84228580-1c14-11ea-8f2a-94ebd2afa14c.png)

> So the question for the spec editors is, do you want to allow clicking on the top-level Window to activate a PaymentRequest object that comes from a same-origin iframe? 

I don't think we do. The activated window should be the one where `request.show()` is called.

For the clarity of those watching at home:

```HTML 
  <button>pay</button>
  <iframe src="same-origin.html" hidden>
    <!-- this iframe won't activate the payment request. --> 
  </iframe> 
```

This will never ever work, which is fine:
```HTML
  <button>pay</button>
  <iframe allowpaymentrequest src="https://cross.origin/" hidden>
  </iframe>
```

This will work: 
```HTML
  <iframe src="same-origin.html">
    <!-- something in iframe is clicked/activated, payment request works --> 
  </iframe>
   <iframe allowpaymentrequest src="https://cross.origin/" hidden>
  </iframe>
``` 

> Personally I'd lean toward the more conservative option, but I'm unsure what folks have implemented, or what kind of workflows the PaymentRequest API hopes to support in this fashion.

I agree. Let's go with that...  or **this** 😂.